lms/djangoapps/experiments/views.py

```diff
@@ -47,3 +54,31 @@ def create_or_update(self, request, *args, **kwargs):
 
         self.action = 'create'
         return self.create(request, *args, **kwargs)
+
+    def _cache_users(self, usernames):
+        users = User.objects.filter(username__in=usernames)
+        self._cached_users = {user.username: user for user in users}
+
+    def _get_user(self, username):
+        user = self._cached_users.get(username)
+
+        if not user:
+            user = User.objects.get(username=username)
+            self._cached_users[username] = user
+
+        return user
+
+    @list_route(methods=['put'], permission_classes=[permissions.IsAdminUser])
+    def bulk_upsert(self, request):
+        upserted = []
+        self._cache_users([datum['user'] for datum in request.data])
+
+        with transaction.atomic():
+            for item in request.data:
+                user = self._get_user(username=item['user'])
+                datum, __ = ExperimentData.objects.update_or_create(
```

Comment:
If we don't batch the SQL queries, can't the response time be too slow and exceed timeouts?